File: Data Structures & Algorithms/binary-tree-preorder-traversal/submission-1.cs
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left;
 *     public TreeNode right;
 *     public TreeNode(int val=0, TreeNode left=null, TreeNode right=null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

public class Solution {
    public List<int> PreorderTraversal(TreeNode root) {
        var stk = new Stack<TreeNode>();
        var ans = new List<int>();
        var cur = root;

        while(stk.Count > 0 || cur != null)
        {
            while(cur != null)
            {
                ans.Add(cur.val);
                stk.Push(cur);
                cur = cur.left;
            }

            stk.TryPop(out cur);
            cur = cur.right;
        }

        return ans;
    }
}

File: Data Structures & Algorithms/count-good-nodes-in-binary-tree/submission-1.ts
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     constructor(val = 0, left = null, right = null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    /**
     * @param {TreeNode} root
     * @return {number}
     */
    goodNodes(root: TreeNode | null): number {
        // track highest node in a var
        // track all nodes in the path in a map
        // use stack for dfs
        if(!root){
            return 0;
        }

        let stk = [];
        let path = []
        let lastVisited = null;
        let highest = [root.val];
        let cur = root;
        let count = 0;

        while(stk.length > 0 || cur){
            while(cur){
                stk.push(cur);
                path.push(cur.val);
                if(cur.val >= highest[highest.length-1]){
                    count++;
                    highest.push(cur.val);
                }
                cur = cur.left;
            }
            let peek = stk[stk.length-1];
            if(peek.right && lastVisited != peek.right){
                cur = peek.right;
            }else {
                let pathval = path.pop();
                if(pathval === highest[highest.length-1]){
                    highest.pop();
                }
                lastVisited = stk.pop()
            }
        }
        return count;
    }
}














